src/OpenLoco/Industry.cpp

```diff
@@ -439,30 +440,179 @@ namespace OpenLoco
-        call(0x00454A43, regs);
+        if (!Map::validCoords(pos))
+        {
+            return false;
+        }
+
+        const auto tile = Map::TileManager::get(pos);
+        bool passedSurface = false;
+        for (auto& el : tile)
+        {
+            auto* elSurface = el.as<Map::SurfaceElement>();
+            if (elSurface != nullptr)
+            {
+                if (elSurface->water() != 0)
+                {
+                    return false;
+                }
+                if (elSurface->hasType6Flag())
+                {
+                    return false;
+                }
+                passedSurface = true;
+                continue;
+            }
+            if (!passedSurface)
+            {
+                continue;
+            }
+            if (el.isGhost())
+            {
+                continue;
+            }
+            if (el.as<Map::WallElement>() != nullptr || el.as<Map::TreeElement>() != nullptr)
+            {
+                continue;
+            }
+            return false;
+        }
+
+        return true;
+    }
+
+    // 0x0045510C bl == 1
+    static bool claimSurfaceForIndustry(const Map::TilePos2& pos, IndustryId industryId, uint8_t var_EA)
+    {
+        if (!isSurfaceClaimed(pos))
+        {
+            return false;
+        }
+
+        const auto tile = Map::TileManager::get(pos);
+        Map::SurfaceElement* surface = tile.surface();
+        surface->setHighTypeFlag(true);
+        surface->setIndustry(industryId);
+        surface->setVar5SLR5((var_EA & 0xE0) >> 5);
+        surface->setVar6SLR5((var_EA & 0x7));
+        Ui::ViewportManager::invalidate(pos, surface->baseHeight(), surface->baseHeight() + 32);
+        Map::TileManager::removeAllWallsOnTile(pos, surface->baseZ());
+
+        return true;
+    }
+
+    // 0x00454A43
+    void Industry::expandGrounds(const Pos2& pos, uint8_t primaryWallType, uint8_t secondaryWallType, uint8_t dl)
+    {
+        std::size_t numBorders = 0;
+        // Search a 5x5 area centred on Pos
+        TilePos2 topRight = TilePos2{ pos } - TilePos2{ 2, 2 };
+        TilePos2 bottomLeft = TilePos2{ pos } + TilePos2{ 2, 2 };
+        for (const auto& tilePos : TilePosRangeView{ topRight, bottomLeft })
+        {
+            if (isSurfaceClaimed(tilePos))
+            {
+                numBorders++;
+            }
+        }
+        if (numBorders < 20)
+        {
+            return;
+        }
+
+        const auto* indObj = ObjectManager::get<IndustryObject>(objectId);
+
+        std::optional<Utility::prng> is23prng;
+        if (indObj->flags & IndustryObjectFlags::unk23) // Livestock use this
+        {
+            is23prng = prng;
+        }
+        std::optional<Utility::prng> is27prng;
+        if (indObj->flags & IndustryObjectFlags::unk27) // Skislope use this
+        {
+            // Vanilla mistake here didn't set the prng! It would just recycle from a previous unk23 caller
+            is27prng = prng;
```

Comment:
Any notable divergence as a result of this change?

Reply:
Yes there would be but I've not really got a good setup to show the difference as its only used by ski centres.